posts/ngrx-facade-pattern.md

```diff
@@ -0,0 +1,137 @@
+* `Selectors` are pure functions that derive and compose a read-only piece of state.
+* We select a piece of state from the `store` using `selectors`.
+
+## Facade Pattern
+Now we will dive into what is the Facade Pattern and how to implement it with NgRx.
+
+### What is a Facade?
+In software, a facade provides a public interface to mask the usage of something more complex. In this instance, it will neatly wrap our NgRx interactions and allow one point of contact for our component and state management without our component having any knowledge of NgRx.
+
+### Implementation
+Before jumping in to the facade example, here is a component using NgRx without a facade.
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private store: Store<AccountListState>) {}
+
+  ngOnInit() {
+    this.accountList = this.store.select(getAccountList);
+
+    this.loadAccountList();
+  }
+
+  loadAccountList() {
+    this.store.dispatch(new LoadAccountList());
+  }
+
+  addAccount(account: string) {
+    this.store.dispatch(new AddAccount(account));
+  }
+
+  editAccount(id: string, account: string) {
+    this.store.dispatch(new EditAccount({ id, account }));
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+This implementation is very `store` dependent and makes the component more complex. By adding a facade for the component to interact with, we can simplify the access to our store and decouple the knowledge of NgRx in our component. By performing that separation of duties, if we choose to remove NgRx in the future we do not have to refactor our component.
+
+Now let's create an `AccountService` that will act as our facade for NgRx.
+``` Typescript
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AccountService {
+  accountList = this.store.select(getAccountList);
+
+  constructor(private store: Store<AccountListState>) {}
+
+  loadAccountList() {
+    this.store.dispatch(new LoadAccountList());
+  }
+
+  addAccount(account: string) {
+    this.store.dispatch(new AddAccount(account));
+  }
+
+  editAccount(id: string, account: string) {
+    this.store.dispatch(new EditAccount({ id, account }));
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+After moving everything related to the store into our service, we can then inject that service into our component.
+
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private accountService: AccountService) {}
+
+  ngOnInit() {
+    this.accountList = this.accountService.accountList;
+
+    this.accountService.loadAccountList();
+  }
+
+  addAccount(account: string) {
+    this.accountService.addAccount(account);
+  }
+
+  editAccount(id: string, account: string) {
+    this.accountService.editAccount({ id, account });
```

Comment:
AccountService takes two params instead of an object.  Would align it one way or the other.

Reply:
@johnstrickler updated. That was supposed to be two separate string params